src/lsm/tree.zig

```diff
@@ -266,6 +272,88 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
+            if (tree.key_min == null or compare_keys(key, tree.key_min.?) == .lt) {
+                tree.key_min = key;
+            }
+            if (tree.key_max == null or compare_keys(key, tree.key_max.?) == .gt) {
+                tree.key_max = key;
+            }
+            assert(tree.key_min != null and tree.key_max != null);
+        }
+
+        /// Returns True if the given key may be present in the Tree/Manifest, False if the key is
+        /// guaranteed to not be present.
+        ///
+        /// Performs the following steps based on the provided location:
+        /// *  tree -> Checks whether the key exists within the Tree's key range.
+        /// *  manifest -> Checks whether the key exists within the ManifestLevels' key range.
+        ///    Then, for each ManifestLevel that could contain the key, it attempts to fetch the
+        ///    index block, filter block & data block for the tables that could contain the key
+        ///    synchronously from the Grid cache. If any of these blocks aren't present in the cache,
+        ///    it returns True since the existence of the key cannot be determined.
+        pub fn key_may_exist(tree: *Tree, key: Key, location: union(enum) { tree, manifest }) bool {
+            assert((tree.key_min == null and tree.key_max == null) or
+                (tree.key_min != null and tree.key_max != null));
+            switch (location) {
+                .tree => {
+                    // Tree key range hasn't been updated yet.
+                    if (tree.key_min == null and tree.key_max == null) {
+                        return false;
+                    }
+                    if (compare_keys(key, tree.key_min.?) == .lt or compare_keys(key, tree.key_max.?) == .gt) {
+                        return false;
+                    } else {
+                        return true;
+                    }
+                },
+                .manifest => {
+                    var iterator = tree.manifest.key_may_exist(key);
+                    if (iterator.next() == null) {
+                        return false;
+                    }
+                    const fingerprint = bloom_filter.Fingerprint.create(stdx.hash_inline(key));
+                    while (iterator.next()) |table| {
+                        if (tree.grid.read_from_cache(table.address, table.checksum)) |index_block| {
+                            const blocks = Table.index_blocks_for_key(index_block, key);
+                            if (tree.grid.read_from_cache(
+                                blocks.filter_block_address,
+                                blocks.filter_block_checksum,
+                            )) |filter_block| {
+                                const filter_bytes = Table.filter_block_filter_const(filter_block);
+                                if (!bloom_filter.may_contain(fingerprint, filter_bytes)) {
+                                    continue;
+                                }
```

Comment:
Let's comment the else branch to be clear about why falling through is safe.

Reply:
It might also be good to extract these two blocks into two functions, so that in `verify` mode we can double check (as far as possible) that if the filter decides to continue, then the key is also checked to not be in the data block.

Reply:
Took care of extracting them out into two functions in https://github.com/tigerbeetle/tigerbeetle/pull/944/commits/daf301b87126e3490197b3c199dcfb5fb070d8f5.